### PR TITLE
Improve table records count calculation in pipeline job for MySQL

### LINKED ISSUE #24291: Improve table records count calculation in pipeline job
### Is your feature request related to a problem?
Yes

### Describe the feature you would like.
Currently, `InventoryTaskSplitter.getTableRecordsCount` use `SELECT COUNT(*) FROM table` to query records count. It might cost too much time, even more than time out of connection.

A possible solution is to query estimated records count from database.

Some investigation on databases:

#### MySQL investigation

Query from `INFORMATION_SCHEMA.TABLES`. Example:

```
mysql> select table_schema, table_name, table_rows from INFORMATION_SCHEMA.TABLES where table_schema='sysbench_ds_1';
+---------------+------------+------------+
| table_schema  | table_name | table_rows |
+---------------+------------+------------+
| sysbench_ds_1 | sbtest1    |  138247749 |
+---------------+------------+------------+
1 row in set (0.02 sec)

mysql> analyze table sysbench_ds_1.sbtest1;
+-----------------------+---------+----------+----------+
| Table                 | Op      | Msg_type | Msg_text |
+-----------------------+---------+----------+----------+
| sysbench_ds_1.sbtest1 | analyze | status   | OK       |
+-----------------------+---------+----------+----------+
1 row in set (0.46 sec)

mysql> select table_schema, table_name, table_rows from INFORMATION_SCHEMA.TABLES where table_schema='sysbench_ds_1';
+---------------+------------+------------+
| table_schema  | table_name | table_rows |
+---------------+------------+------------+
| sysbench_ds_1 | sbtest1    |  138082248 |
+---------------+------------+------------+
1 row in set (0.02 sec)

mysql> explain select count(1) from sysbench_ds_1.sbtest1;
+----+-------------+---------+------------+-------+---------------+------+---------+------+-----------+----------+-------------+
| id | select_type | table   | partitions | type  | possible_keys | key  | key_len | ref  | rows      | filtered | Extra       |
+----+-------------+---------+------------+-------+---------------+------+---------+------+-----------+----------+-------------+
|  1 | SIMPLE      | sbtest1 | NULL       | index | NULL          | k_1  | 4       | NULL | 138082248 |   100.00 | Using index |
+----+-------------+---------+------------+-------+---------------+------+---------+------+-----------+----------+-------------+
1 row in set, 1 warning (0.16 sec)

mysql> select count(1) from sysbench_ds_1.sbtest1;
+-----------+
| count(1)  |
+-----------+
| 140000000 |
+-----------+
1 row in set (36.82 sec)
```

`0.02 sec` vs `36.82 sec` on 140 millions records table. The estimated records count has accurateness lost: `(140000000-138082248)/140000000.0*100 ~= 1.37`, it's acceptable.

#### PostgreSQL investigation

Query from `pg_class`. Example:
```
test1=# select count(1) from t_order;
 count
-------
     6
(1 row)

test1=# select relname,reltuples::integer from pg_class where relname='t_order';
 relname | reltuples
---------+-----------
 t_order |         0
(1 row)


test1=# analyze t_order;
ANALYZE

test1=# select relname,reltuples::integer from pg_class where relname='t_order';
 relname | reltuples
---------+-----------
 t_order |         6
(1 row)
```

TODO: it needs more test on large table.

#### Possible changes

- Add method `Optional<String> buildEstimatedCountSQL` in `PipelineSQLBuilder`, implement it if database support records count estimation. We could support `MySQL`, `PostgreSQL` and `openGauss` for now.
- If estatimated records count is not supported, then use `buildCountSQL`

Notice:
- The estimated records count might return `0` from database

### PR DESCRIPTION
**Pull request by azexcy**
Fixes #24291.

Changes proposed in this pull request:
  - Improve table records count calculation in pipeline job

---

Before committing this PR, I'm sure that I have checked the following options:
- [ ] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [ ] I have self-reviewed the commit code.
- [ ] I have (or in comment I request) added corresponding labels for the pull request.
- [ ] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [ ] I have made corresponding changes to the documentation.
- [ ] I have added corresponding unit tests for my changes.

### CODE REVIEW

File: kernel/data-pipeline/api/src/main/java/org/apache/shardingsphere/data/pipeline/spi/sqlbuilder/PipelineSQLBuilder.java [hunk 1]
```diff
@@ -135,6 +135,16 @@ default Optional<String> buildCreateSchemaSQL(String schemaName) {
...
     
+    /**
+     * Build Estimate count SQL.
+     *
+     * @param databaseName database name
+     * @param schemaName schema name
+     * @param tableName table name
+     * @return estimate count sql
+     */
+    String buildEstimateCountSQL(String databaseName, String schemaName, String tableName);
```
Comment (sandynz) [inline kernel/data-pipeline/api/src/main/java/org/apache/shardingsphere/data/pipeline/spi/sqlbuilder/PipelineSQLBuilder.java:146]: `buildEstimateCountSQL` could be `buildEstimatedCountSQL`

File: kernel/data-pipeline/core/src/main/java/org/apache/shardingsphere/data/pipeline/core/prepare/InventoryTaskSplitter.java [hunk 2]
```diff
@@ -171,12 +171,20 @@ private long getTableRecordsCount(final InventoryIncrementalJobItemContext jobIt
...
         String schemaName = dumperConfig.getSchemaName(new LogicTableName(dumperConfig.getLogicTableName()));
         String actualTableName = dumperConfig.getActualTableName();
-        // TODO with a large amount of data, count the full table will have performance problem
-        String sql = PipelineTypedSPILoader.getDatabaseTypedService(PipelineSQLBuilder.class, jobConfig.getSourceDatabaseType()).buildCountSQL(schemaName, actualTableName);
+        PipelineSQLBuilder pipelineSQLBuilder = PipelineTypedSPILoader.getDatabaseTypedService(PipelineSQLBuilder.class, jobConfig.getSourceDatabaseType());
+        String sql = pipelineSQLBuilder.buildEstimateCountSQL(dumperConfig.getDataSourceName(), schemaName, actualTableName);
         try (
                 Connection connection = dataSource.getConnection();
                 PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
+            long estimateCount;
```
Comment (sandynz) [inline kernel/data-pipeline/core/src/main/java/org/apache/shardingsphere/data/pipeline/core/prepare/InventoryTaskSplitter.java:179]: `estimateCount` could be `result`

File: kernel/data-pipeline/core/src/main/java/org/apache/shardingsphere/data/pipeline/core/prepare/InventoryTaskSplitter.java [hunk 3]
```diff
@@ -171,12 +171,20 @@ private long getTableRecordsCount(final InventoryIncrementalJobItemContext jobIt
...
             try (ResultSet resultSet = preparedStatement.executeQuery()) {
+                resultSet.next();
+                estimateCount = resultSet.getLong(1);
+            }
+            if (estimateCount > 0) {
+                return estimateCount;
+            }
+            try (ResultSet resultSet = connection.createStatement().executeQuery(pipelineSQLBuilder.buildCountSQL(schemaName, actualTableName))) {
```
Comment (sandynz) [inline kernel/data-pipeline/core/src/main/java/org/apache/shardingsphere/data/pipeline/core/prepare/InventoryTaskSplitter.java:187]: 1, `buildCountSQL` might cost too much time, so add some log for it, includes cost time.

2, connection will return to pool but not closed, so will `connection.createStatement()` be closed automatically?

File: kernel/data-pipeline/dialect/mysql/src/main/java/org/apache/shardingsphere/data/pipeline/mysql/sqlbuilder/MySQLPipelineSQLBuilder.java [hunk 4]
```diff
@@ -89,6 +89,11 @@ public Optional<String> buildCRC32SQL(final String schemaName, final String tabl
         return Optional.of(String.format("SELECT BIT_XOR(CAST(CRC32(%s) AS UNSIGNED)) AS checksum, COUNT(1) AS cnt FROM %s", quote(column), quote(tableName)));
     }
     
+    @Override
+    public String buildEstimateCountSQL(final String databaseName, final String schemaName, final String tableName) {
+        return String.format("SELECT TABLE_ROWS from INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA = '%s' AND TABLE_NAME = '%s'", databaseName, getQualifiedTableName(schemaName, tableName));
```
Comment (sandynz) [inline kernel/data-pipeline/dialect/mysql/src/main/java/org/apache/shardingsphere/data/pipeline/mysql/sqlbuilder/MySQLPipelineSQLBuilder.java:94]: `from` could be `FROM`

File: kernel/data-pipeline/dialect/postgresql/src/main/java/org/apache/shardingsphere/data/pipeline/postgresql/sqlbuilder/PostgreSQLPipelineSQLBuilder.java [hunk 5]
```diff
@@ -85,6 +85,12 @@ private String buildConflictSQL(final DataRecord dataRecord) {
         return result.toString();
     }
     
+    @Override
+    public String buildEstimateCountSQL(final String databaseName, final String schemaName, final String tableName) {
+        // TODO Support estimate count later.
+        return buildCountSQL(schemaName, tableName);
+    }
```
Comment (sandynz) [inline kernel/data-pipeline/dialect/postgresql/src/main/java/org/apache/shardingsphere/data/pipeline/postgresql/sqlbuilder/PostgreSQLPipelineSQLBuilder.java:92]: It's not supported, so it should return `Optional<String>`

**Review by sandynz — CHANGES_REQUESTED** [review]

File: kernel/data-pipeline/dialect/postgresql/src/main/java/org/apache/shardingsphere/data/pipeline/postgresql/sqlbuilder/PostgreSQLPipelineSQLBuilder.java (hunk 5 above)
Comment (azexcy) [reply kernel/data-pipeline/dialect/postgresql/src/main/java/org/apache/shardingsphere/data/pipeline/postgresql/sqlbuilder/PostgreSQLPipelineSQLBuilder.java:92]: Improved

File: kernel/data-pipeline/core/src/main/java/org/apache/shardingsphere/data/pipeline/core/prepare/InventoryTaskSplitter.java [hunk 6]
```diff
@@ -171,21 +172,48 @@ private long getTableRecordsCount(final InventoryIncrementalJobItemContext jobIt
         PipelineJobConfiguration jobConfig = jobItemContext.getJobConfig();
         String schemaName = dumperConfig.getSchemaName(new LogicTableName(dumperConfig.getLogicTableName()));
         String actualTableName = dumperConfig.getActualTableName();
-        // TODO with a large amount of data, count the full table will have performance problem
-        String sql = PipelineTypedSPILoader.getDatabaseTypedService(PipelineSQLBuilder.class, jobConfig.getSourceDatabaseType()).buildCountSQL(schemaName, actualTableName);
+        PipelineSQLBuilder pipelineSQLBuilder = PipelineTypedSPILoader.getDatabaseTypedService(PipelineSQLBuilder.class, jobConfig.getSourceDatabaseType());
+        Optional<String> estimatedCountSQL = pipelineSQLBuilder.buildEstimatedCountSQL(schemaName, actualTableName);
```
Comment (sandynz) [inline kernel/data-pipeline/core/src/main/java/org/apache/shardingsphere/data/pipeline/core/prepare/InventoryTaskSplitter.java:176]: `estimatedCountSQL` could be `sql`

File: kernel/data-pipeline/core/src/main/java/org/apache/shardingsphere/data/pipeline/core/prepare/InventoryTaskSplitter.java [hunk 7]
```diff
@@ -171,21 +172,48 @@ private long getTableRecordsCount(final InventoryIncrementalJobItemContext jobIt
...
+            }
+            return getCountSQLResult(dataSource, pipelineSQLBuilder.buildCountSQL(schemaName, actualTableName));
+        } catch (final SQLException ex) {
+            String uniqueKey = dumperConfig.hasUniqueKey() ? dumperConfig.getUniqueKeyColumns().get(0).getName() : "";
+            throw new SplitPipelineJobByUniqueKeyException(dumperConfig.getActualTableName(), uniqueKey, ex);
+        }
+    }
+    
+    // TODO maybe need refactor after PostgreSQL support estimated count.
+    private long getEstimatedCountSQLResult(final DataSource dataSource, final String estimatedCountSQL) throws SQLException {
```
Comment (sandynz) [inline kernel/data-pipeline/core/src/main/java/org/apache/shardingsphere/data/pipeline/core/prepare/InventoryTaskSplitter.java:190]: `getEstimatedCountSQLResult` could be `getEstimatedCount`.

And also `getCountSQLResult` could be `getCount`.

**Review by sandynz — CHANGES_REQUESTED** [review]

**Review by sandynz — APPROVED** [review]

**Pull request merged**